### ISSUE TITLE
Overlap

### ISSUE THREAD

**Issue by tbreloff** [JuliaML/Losses.jl#7]
I'm starting this as a catch-all for discussion about concepts that heavily overlap between disciplines.  I had started to contemplate how to merge the implementation that you're developing here with the concepts/requirements that I see in OnlineStats/OnlineAI.

(sorry... this is a little rambling...)

Similar:
- Link functions in GLM (https://en.wikipedia.org/wiki/Generalized_linear_model#Link_function)
- Activation functions in ANNs (https://en.wikipedia.org/wiki/Activation_function)
- Kernels (could be considered similar with the right perspective)

A link function is effectively of the form:  `f(y) = x'w + b` or `y = f'(x'w + b)`
An activation function is effectively:  `y = f(x'w + b)`... the name is more a holdover from the perceptron's discontinuous {0,1} classification, but over time has seemed to blend more with the concept of a link function.  So we see a "sigmoid activation" is effectively the same as the "inverse logit link" in the GLM sense.  It's valuable to think of a neural net with no hidden layers to be conceptually similar to a GLM... we are transforming the inner product of inputs and weights (plus bias), with the goal of maximizing the likelihood that w/b generate the relationship.

Now, neural nets are built from connected layers of these GLMs, which just happen to have solvers that exploit the known structure to provide better "measures of fit" for the earlier layers.  The likelihood is in relation to the whole of the structure, as opposed to just a single function of `f(x'w + b)`.  

Lets jump ahead and think of how we can apply a "layering" concept to other areas.  You mentioned in another issue that you would like to do something like:
```
cs = fit(CenterScale(), Xtrain)
Xtrain2 = predict(cs, Xtrain)

pca = fit(PCA(), Xtrain2)
Xtrain3 = predict(pca, Xtrain2)

myfit = fit(MyModel(), Xtrain3)
yhat = predict(myfit, Xtrain3)
```

In some sense this is an extremely specific 3-layer neural net, and I think some abstractions can be seen.  Assume we have 3 input variables, and 1 target.  Imagine instead:
- the first layer is `y1 = f1(x'w1 + b1)`, with w1 the identity matrix, and f1 the identity function.
- the second layer is `y2 = f2(y1'w2 + b2)`, with w2 as the principal components loading matrix, f2 the identity function
- the third later is the model `y3 = f3(y2, w3, b3)`

Thinking about things like this I think will open your mind to some alternative ways to think about modeling pipelines... specifically why are you focused on de-meaning the inputs, or reducing dimensions using PCA, if your final goal is really to have some larger function which maps initial inputs to final target values.  The concept behind [PLS](http://www.ats.ucla.edu/stat/sas/library/pls.pdf) is very similar... instead of 2 independent steps of "reduce dimension, then fit", it combines into a unified step "reduce dimensions and fit".  This is how neural nets work as well, and backprop just gives a method to fit the weights of earlier models using the chain rule and intermediate gradients.

There are also parallels to a lot of the deep learning approaches... namely the use of Restricted Boltzmann Machines and Stacked Autoencoders (both have heavy parallels with PCA and dimensionality reduction) as inputs into a simpler predictive model.  There has been good success in using these models/algorithms simply as weight initialization in a larger system, where those models are effectively re-fit using the final target as driving the weight updates, here with a heavy parallel to PLS.

The point I'm trying to get across with all these tangents is that I think it can be very helpful when designing, to think about how simpler models (pca, svm, glm, ...) could be reinterpreted in terms of a more complicated framework.  With the right abstractions, we could recreate something like PLS automatically by plugging the generalized PCA with generalized logistic regression.

**Comment by Evizero** [comment]
I agree with the goal of creating this toolbox of lego pieces that nicely fit and can form bigger things that can again be treated as one thing etc. I think this should be an explicit focus of us

I get why you use neural networks as example, but to me neural networks come attached with the concept of propagating something backwards as well. I think it is more natural to think of it as a data science pipeline. Especially when it comes to terminology (e.g. having a OneHotEncodingLayer for example just feels forced)

A core issue that is very important to me is that the pieces are useful for teaching and research as well. That means that they have to be flexible to allow for targeted experiments and make sense to the domain experts. So the small pieces have to be useful on their own. A loss function should be useful to someone who wants to mess around with loss functions. SVMs are a good example of how complex that can be. "Just" having the ability to fit some pre specified SVMs like `SVM(kernel="rbf", loss="l2")` is not good enough.

My point is that it's difficult to plan such a complex ecosystem in advance. At some point we have to start the coding and iterate on it. I'm in the process of refactoring KSVM.jl to be based on LearnBase

**Comment by tbreloff** [comment]
> but to me neural networks come attached with the concept of propagating something backwards as well

Looking at classic neural nets, I agree that they go hand in hand with "propagating something backwards", but that doesn't need to be the building block.  There are many pieces to the neural net that heavily overlap with other techniques.

> My point is that it's difficult to plan such a complex ecosystem in advance

Total agreement, which is why I keep preaching to stop focusing on the type hierarchy and focus more on verbs and actions.  We should use duck-typing exclusively, except for a few very general abstracts just to classify which verbs might apply.  Who cares if we copy some code initially... at least it's easy to read.

For example, we don't need [this](https://github.com/Evizero/LearnBase.jl/blob/master/src/loss/abstract_cost.jl#L135-L199) when we are going to define [this](https://github.com/Evizero/LearnBase.jl/blob/master/src/loss/margin_based.jl#L21-L35) explicitly later.  In my mind, it's just adding to the codebase and complexity without much benefit.  

Here's some possibilities for those few abstracts:

- `Model`: Maps input to output... minimum amount of information to represent `f(x'w + b)`, for example
  - Doesn't imply size or shape (or type) of input or output data
- `Algorithm`: Controls the `fit` method of the model
  - here we can control things like: online vs batch, independent vs pipelined/layered, etc.
  - I also think we can define concepts like backprop here... doesn't need to be part of the `Model`
- `Loss`:  Defines loss function and derivatives (but not gradients?)
  - can be subtype of `Cost`, but that (maybe) only matters if we intend to merge with the generic optimization frameworks somday
- `Penalty`: additional cost as function of fitted parameters
  - I actually go back and forth on whether this should be part of `Cost`... i.e. there could be a `Cost` which is the sum of a `Loss` and a `Penalty`

So generally speaking, a neural net layer is defined by a `Model` which combines an `LinearCombination` with a `NonlinearMap` or something like that.  We never need to define `type NeuralNetLayer <: Model`, although we may choose to do that for other reasons.  Then the neural net is just a vector of these layers.  A `BackProp <: Algorithm` could make forward and backward passes over this vector of `Model`s, and this is the only part that necessarily knows we are fitting a neural net.

I understand it's probably hard to visualize how the whole ecosystem fits together... all the more reason to keep the code minimal and sparse.  I forked the repo, and I'll start to build up some of the building blocks using this concept and see how far I get.  I think that if we can abstract this list into the same verbs/framework then we'll be in good shape to extend to other techniques:

- PCA (unsupervised learning)
- SVM (margin-based classification)
- Multinomal probabilities from neural net

**Comment by tbreloff** [comment]
I'm also thinking through whether we can depend on https://github.com/mauro3/SimpleTraits.jl and prepare for traits to eventually be part of core Julia.

**Comment by Evizero** [comment]
I agree with pretty much everything expect the following

> For example, we don't need this when we are going to define this explicitly later. In my mind, it's just adding to the codebase and complexity without much benefit. 

I very strongly disagree (aside from the stubs that have the "not implemented" macro). The MarginBased vs DistanceBased loss is the most useful aspect of the Loss functions to me. I absolutely need those in some form for KSVM. Most importantly they allow dispatch on Classification vs Regression on the loss level. Also I don't want the loss functions to be some invisible backbone stuff that shouldn't be touched by endusers. I would like them to be useful on their own for course exercises and experiments. I also need the properties. For example the Pegasos algorithm for SVMs needs the loss to be lipschitzcontinuous.

> ... Defines loss function and derivatives (but not gradients?) 

Well, practically speaking the difference is the type of the parameters. might as well use one verb for both. That being said, I have no strong opinion on this.

> I actually go back and forth on whether this should be part of Cost... i.e. there could be a Cost which is the sum of a Loss and a Penalty

I'm also not completely set on this. Initially loss+penalty was the cost to me, but that would be inconsistent with the loss theory. I'll soon sketch out deterministic empirical risk minimization using Optim.jl and see where it takes me

>  I forked the repo, and I'll start to build up some of the building blocks using this concept and see how far I get.

sweet! I'll continue working on loss/predictor as I progress on KSVM

**Comment by Evizero** [comment]
> ... go back and forth on whether this should be part of Cost

I do think though that a Penalty should exist on their own. So that I can have different parameters for `Loss` and `Penalty`, which is important

**Comment by Evizero** [comment]
> I'm also thinking through whether we can depend on https://github.com/mauro3/SimpleTraits.jl and prepare for traits to eventually be part of core Julia.

I'm open to it if it turns out to be helpful to do something

**Comment by tbreloff** [comment]
> The MarginBased vs DistanceBased loss is the most useful aspect of the Loss functions to me

It might help me understand if I see it in use.  Do you have any integration with your KSVM package (or any other examples) that you can show my how it's used?

**Comment by tbreloff** [comment]
Here's an example from within SimpleTraits, that might give an idea of one way we could use traits.
```
"Trait of all leaf types types"
@traitdef IsLeafType{X}
@generated trait{X}(::Type{IsLeafType{X}}) = X.mutable ? :(Not{IsLeafType{X}}) : :(IsLeafType{X})
```

We could maybe use your "is..." methods (like `istwicedifferentiable(::LogitMarginLoss) = true`) to define a series of traits that can uniquely define the generic methods that apply.  So for this example, maybe:

```
@traitdef IsTwiceDifferentiable{X}
@generated trait{X}(::Type{IsTwiceDifferentiable{X}}) = if istwicedifferentiable(X)
  :(IsTwiceDifferentiable{X})
else
  :(Not{IsTwiceDifferentiable{X}})
end
```

and it would allow dispatch on the properties of the type, not the type hierarchy.
